### PR TITLE
292 tm process aos process oob read and sa not operational

### PR DESCRIPTION
**Pull request by Donnie-Ice**

### CODE REVIEW

File: src/core/crypto_aos.c [hunk 1]
```diff
@@ -1249,6 +1249,13 @@ int32_t Crypto_AOS_ProcessSecurity(uint8_t* p_ingest, uint16_t len_ingest, uint8
 
     }
 
+    // check sa state before decrypting
+    if (sa_ptr->sa_state != SA_OPERATIONAL)
+    {
+        printf(KRED "Error: SA Not Operational \n" RESET);
```
Comment (rjbrown2) [inline src/core/crypto_aos.c:1255]: Can you wrap these in #ifdef DEBUG cases?  Any print we supply should only be used in Debug cases.

File: src/core/crypto_tm.c [hunk 2]
```diff
@@ -1787,6 +1787,12 @@ int32_t Crypto_TM_ProcessSecurity(uint8_t* p_ingest, uint16_t len_ingest, uint8_
         // Parse MAC, prepare AAD
         Crypto_TM_Parse_Mac_Prep_AAD(sa_service_type, p_ingest, mac_loc, sa_ptr, &aad_len, byte_idx, aad);
 
+        if (sa_ptr->sa_state != SA_OPERATIONAL)
+        {
+            printf(KRED "Error: SA Not Operational \n" RESET);
```
Comment (rjbrown2) [inline src/core/crypto_tm.c:1792]: #ifdef DEBUG

File: src/sa/internal/sa_interface_inmemory.template.c [hunk 3]
```diff
@@ -980,13 +980,13 @@ static int32_t sa_start(TC_t* tc_frame)
...
-    // Overwrite last PID
-    sa[spi].lpid =
-        (sdls_frame.pdu.type << 7) | (sdls_frame.pdu.uf << 6) | (sdls_frame.pdu.sg << 4) | sdls_frame.pdu.pid;
-
     // Check SPI exists and in 'Keyed' state
     if (spi < NUM_SA)
     {
+        // Overwrite last PID
+        sa[spi].lpid =
+            (sdls_frame.pdu.type << 7) | (sdls_frame.pdu.uf << 6) | (sdls_frame.pdu.sg << 4) | sdls_frame.pdu.pid;
```
Comment (rjbrown2) [inline src/sa/internal/sa_interface_inmemory.template.c:988]: Can we add a comment here for what is actually going on here?

File: src/sa/internal/sa_interface_inmemory.template.c [hunk 4]
```diff
@@ -1109,13 +1109,13 @@ static int32_t sa_stop(void)
...
-    // Overwrite last PID
-    sa[spi].lpid =
-        (sdls_frame.pdu.type << 7) | (sdls_frame.pdu.uf << 6) | (sdls_frame.pdu.sg << 4) | sdls_frame.pdu.pid;
-
     // Check SPI exists and in 'Active' state
     if (spi < NUM_SA)
     {
+        // Overwrite last PID
+        sa[spi].lpid =
+            (sdls_frame.pdu.type << 7) | (sdls_frame.pdu.uf << 6) | (sdls_frame.pdu.sg << 4) | sdls_frame.pdu.pid;
```
Comment (rjbrown2) [inline src/sa/internal/sa_interface_inmemory.template.c:1117]: Comment for what shifting is happening here?

File: src/sa/internal/sa_interface_inmemory.template.c [hunk 5]
```diff
@@ -1290,57 +1290,65 @@ static int32_t sa_create(void)
...
+        sa[spi].sa_state = SA_UNKEYED;
 
 #ifdef PDU_DEBUG
-    Crypto_saPrint(&sa[spi]);
+        Crypto_saPrint(&sa[spi]);
 #endif
+    }
+    else
+    {
+        printf(KRED "ERROR: SPI %d cannot be created.\n" RESET, spi);
```
Comment (rjbrown2) [inline src/sa/internal/sa_interface_inmemory.template.c:1350]: #ifdef DEBUG

File: src/sa/internal/sa_interface_inmemory.template.c [hunk 6]
```diff
@@ -1358,13 +1366,13 @@ static int32_t sa_delete(void)
     spi = ((uint8_t)sdls_frame.pdu.data[0] << 8) | (uint8_t)sdls_frame.pdu.data[1];
     printf("spi = %d \n", spi);
```
Comment (rjbrown2) [inline src/sa/internal/sa_interface_inmemory.template.c:1367]: #ifdef DEBUG

File: src/sa/internal/sa_interface_inmemory.template.c [hunk 7]
```diff
@@ -1248,13 +1248,13 @@ static int32_t sa_expire(void)
...
```
Comment (rjbrown2) [inline src/sa/internal/sa_interface_inmemory.template.c:1249]: #ifdef DEBUG here

File: src/sa/internal/sa_interface_inmemory.template.c [hunk 8]
```diff
@@ -1290,57 +1290,65 @@ static int32_t sa_create(void)
...
```
Comment (rjbrown2) [inline src/sa/internal/sa_interface_inmemory.template.c:1291]: #ifdef DEBUG

**Review by rjbrown2 — APPROVED** [review]

**Review by jlucas9 — APPROVED** [review]

**Pull request merged**